tracer/test/Datadog.Trace.Tests/Configuration/ImmutableAzureAppServiceSettingsTests.cs

```diff
@@ -193,5 +194,51 @@ public void NeedsDogStatsD(string value, bool expected)
 
             settings.NeedsDogStatsD.Should().Be(expected);
         }
+
+        [Fact]
+        public void GetIsAzureConsumptionPlanFunctionTrueWhenFunctionsEnvVarsExist()
+        {
+            Environment.SetEnvironmentVariable(ConfigurationKeys.AzureAppService.FunctionsWorkerRuntimeKey, "dotnet");
```

Comment:
(Sorry I didn't get as far as the tests in previews reviews)

Since some test collections can run in parallel, as a general rule we try not to rely on static state like env vars when testing, since this can be the source of flaky tests.

If we add these GCP and AzFunc env vars to the tracer's configuration system, you can test them by using an in-memory collection of key/value pairs instead of using real env vars.

@andrewlock do you think that `Environment.SetEnvironmentVariable()` here is fine since multiple tests from this file won't run at the same time and step on each other's toes?

Reply:
> since multiple tests from the same class won't run at the same time?

I just noticed the same env vars are set from different files. For example, `ConfigurationKeys.GCPFunction.DeprecatedFunctionNameKey`.

Reply:
I think it's unnecessary for those tests in `ServerlessMiniAgentTests` to be accessing the environment (and a bit redundant to have both deprecated & newer GCP environment tests given we test that in `ImmutableGCPFunctionSettingsTests)`. 

I'll change this to specify IsRunningInGCPFunctions when creating the configuration source

Reply:
I can modify the `ServerlessMiniAgentTests.cs` tests to use `ConfigurationKeys` instead, adding `GCPFunctionContextKey` and `AzureFunctionConsumptionPlanContextKey` to their respective configuration key files that would only be used for testing.

Reply:
> @andrewlock do you think that Environment.SetEnvironmentVariable() here is fine since multiple tests from the same class won't run at the same time?

Almost certainly not 😅 We need to be _very_ careful about using global state like this (i.e. basically never do it if you can help it). We have a few places which do, and we're careful to "reset" tests back to their original value afterwards.

The trouble is that setting environment variables like this can "leak" into other tests, changing behaviours (not to mention when running locally)!

Likely the "correct" way to test this is to not use the global static state and/or environment variables at all. Instead, create a new instance of `ImmutableAzureAppServiceSettings`, set the variables in the `IConfigurationSource`, and test the properties of the settings object. You can see an example of this in the tests earlier in this file.

Related to that, `ImmutableAzureAppServiceSettings.GetIsAzureConsumptionPlanFunction()` should likely _not_ be a static function. Instead it should work the same way as the other properties of `ImmutableAzureAppServiceSettings`, i.e. read from the `IConfigurationSource` on creating a new `ImmutableAzureAppServiceSettings` instance and set some properties on the object. These can then be accessed from the instance on `ImmutableTracerSettings`, and as a bonus makes testing safer and easier 🙂

Reply:
> `ImmutableAzureAppServiceSettings.GetIsAzureConsumptionPlanFunction()` should likely not be a static function

Yeah, about that 😅...  I thought about suggesting this as well, but we do have a few specific places where we read env vars directly instead of using a config source. I was thinking it could be fine here since these will always be env vars (they are set by the environment, they are not really user settings).

But even if we stick to env vars only, then we probably want to move it somewhere else instead of a `Immutable*Settings`, since the name alone implied it uses a config source.

I could go either way:
- go all-in using the tracer configuration system (`IConfigurationSource`, etc)
- read env vars directly, but move the methods out of the `Configuration` namespace and rename the type

Reply:
got it, that definitely sounds a lot more robust. I'll make those changes